### ISSUE TITLE
Concurrent script queuing

### ISSUE THREAD

**Issue by ethanopp** [Collisionc/sickbeard_mp4_automator#12]
Is there anyway to ensure the script is only processing 1 video at a time? ...Perhaps a queue?

When I have 2-3 videos finish downloading around the same time, it seems like the script runs in parallel and is hogging a lot of CPU - especially for 2160p downloads

**Comment by bdt1995** [comment]
What client are you using to download? I'm guessing it's torrent client, as the usenet clients, both Sabnzbd and NZBGet run them like you ask, and it's annoying. I got around this by making a .bat that simply includes start <name of sabnzbd script> %1 %2 %3.... (basically just using the batch script as a means of launching it in an externalized process and passing all the arguments to it, so sab doesn't wait for it to finish)

Downside is of course that if more than 2 are running, they fail out since nvenc only supports 2 simulatenous encodes. If you aren't using nvenc, not sure why you're using this fork, since the purpose of this fork of the main repo is nvenc support. I use nvenc, and it doesn't really even touch my cpu, so I wonder if there might be more to your issue than meets the eye. Such as, for instance, unpacking/repairing usenet downloads can eat a huge amount of cpu. Sab has made this less of an issue with Direct Unpack.

It's not hard for **you** to make a queue, since the author likely won't, as it's not central to the core functionality, or even useful to most people. 

The way I'd do it is to set an environmental variable (or a text file, choose your favorite) to 0, and after checking it's 0 and it is, set it to 1 whenever starting a conversion, if it's not 0, run a while loop checking for 0 again every so many seconds (I'd do 15, but that's because my transcodes finish fast, if it's slow you might do up to 60 or even more), then finally when the while loop finishes (it finds the 0), set it to 1, and run your script. Don't forget to pass all your arguments.

If you don't know how to code, this is a good opportunity to learn. Codeacademy.com is wonderful, I recommend python, and you'll have a useful skill. It's also not difficult, to learn, or to do. It always bothers me when people will sit on reddit or stackexchange and ask people to make a script for them, or change their program to fit their specific usecase, when someone with beginner level skills could make a simple script to do what they need. There seems to be this misconception that coding is super complicated and inolves typing in just 1's and 0's. That'd be insanely hard. This isn't about you, as you're not asking someone to do anything, it's just a point I want to make to (hopefully) make another point. Empower yourself. If you already know basic scripting, great! Then you should be able to take my pseudocode and make your own "queue" that'll work for you. 

If not, then maybe hit up codeacademy, and try out the lessons on python. I'm not sure if they have batch specifically, but a quick google search shows they have "Learn the command line", which since batch is merely scripted command prompt commands, I'd say so (batch is even simpler than python, but much less polished and less diverse in its capabilities), but once you learn the basics of coding, you can use most languages, though you tend to have favorites. 

Personally, I'd make this in batch, or PowerShell, but that's me. And that's assuming that for whatever reason the script is running multiple instances in parallel and taking up CPU, and not something else going on. Don't forget to answer my questions about client and such, that's important to figuring out what's happening and how to help.

Hope this helps you.

**Comment by Collisionc** [comment]
Yeah, what @bdt1995 said -- Or suggest adding a queue at https://github.com/mdhiggins/sickbeard_mp4_automator  :) -- which I would eventually merge into this fork.

**Comment by Collisionc** [comment]
Looks like someone already made a pull request for a queue - https://github.com/mdhiggins/sickbeard_mp4_automator/pull/326 

I'll try merging that into my fork and see if it breaks anything after work today.

**Comment by Collisionc** [comment]
Nevermind, that won't do what you're looking for, that will only function with manual.pyu

**Comment by Collisionc** [comment]
I'll go ahead and merge it in, it won't hurt anything, but you'll have to use manual.py to process things instead of radarr.py or whatever you are using.

**Comment by Collisionc** [comment]
https://github.com/Collisionc/sickbeard_mp4_automator/commit/98ab239147ce0414b54703d76427af7a0209a7fa 

Ok merged it in, if you use manual.py add -mp 2 to the arguments and it should limit postprocessing to 2, but only when used with manual.py.

**Comment by ethanopp** [comment]
@Anzo would it work with transmission?

**Comment by Azno** [comment]
It should do as it isn’t Param based

**Comment by Azno** [comment]
@ethanopp if you want to test it with transmission ive uploaded the pull request to mdhiggins (#965) request

**Comment by Collisionc** [comment]
I'll check it out and merge it tomorrow

**Comment by Azno** [comment]
Take it from my fork instead, turned it into a True/False setting

**Comment by ethanopp** [comment]
Thanks guys - sorry am away for holidays so won’t be able to take a look and test until new year

**Comment by Collisionc** [comment]
Thanks @Azno, I merged it in here - https://github.com/Collisionc/sickbeard_mp4_automator/commit/ff9cffa76eac30fed3baa650c3f4401c2ff74e31
block-concurrent=True   will enable this option.

**Comment by ethanopp** [comment]
Just so I’m clear, does this queue them or just cancel one if another is running?

**Comment by Azno** [comment]
It queues them for you

**Comment by Collisionc** [comment]
@Azno  I had an issue with crashing due to block-concurrent not being given a default state.

Fixed here if you want to copy/paste the 1 line change: https://github.com/Collisionc/sickbeard_mp4_automator/commit/ad56ee7e9de2b970de7b4990798bdc340d3279c1#diff-16f000a970304397c3d68b8a52ef543e

**Comment by Azno** [comment]
Sorry, I knew I forgot something, I was doing it via copy and pasting as I didn’t wanna install git on my pc lol... thanks
